### PR TITLE
HPCC-12987 Added case statement for new thor exit codes

### PR DESCRIPTION
**Pull request by Michael-Gardner**
Case statement to catch and act on different exit codes.  Also added formatting fixes to keep the script in line with all the others.  Only issue I see as a possible problem is case 5.  Not sure if that's the correct way to handle TEC_Swap.  @jakesmith 

Signed-off-by: Michael Gardner Michael.Gardner@lexisnexis.com

### CODE REVIEW

File: initfiles/componentfiles/thor/run_thor [hunk 1]
```diff
@@ -22,46 +22,46 @@ thorpid=0
...
-  while ps h $oldpid ; do 
-     echo waiting for process $oldpid to finish
-     sleep 5
-  done
+if [[ -e $RUN_THOR_PID_NAME ]]; then
+    oldpid=$(cat $RUN_THOR_PID_NAME)
+    while ps h $oldpid ; do
+        echo "waiting for process $oldpid to finish"
+        sleep 5
+    done
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: Minor: slightly easier to review, if cosmetic space/formatting changes are split in to a separate commit (but same PR).

File: initfiles/componentfiles/thor/run_thor [hunk 2]
```diff
@@ -69,41 +69,65 @@ while [ 1 ]; do
...
-            echo thor exited cleanly - script stopping
-            exit 0
-        fi
-        echo master exited with errorcode = $errcode
-        if [ ! -e $SENTINEL ]; then
-            echo $SENTINEL 'has been removed (1) - script stopping'
+        case $errcode in
+        # TEC_Clean
+        0)  echo "Thormaster ($thorpid) exited cleanly"
             $deploydir/stop_thor $deploydir
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: This is new, if Thor exited cleanly, then it will have shutdown slaves and stop_thor should not be necessary.

File: initfiles/componentfiles/thor/run_thor [hunk 3]
```diff
@@ -69,41 +69,65 @@ while [ 1 ]; do
...
-            compname=`basename $PWD`
-            swapnode auto $DALISERVER $compname
-            errcode=$?
-            if [ 0 != ${errcode} ]; then
-                echo auto swap node failed, errcode=${errcode}
+            ;;
+        # TEC_CtrlC
+        1)  echo "Thormaster ($thorpid) Interrupted, Ctrl-C caught"
+            $deploydir/stop_thor $deploydir
+            exit 1
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: stop_thor is kill / SIGINT so this is going to be the 'normal' stopping process, so I think this should return exit 0 for now (as it did before).

File: initfiles/componentfiles/thor/run_thor [hunk 4]
```diff
@@ -69,41 +69,65 @@ while [ 1 ]; do
...
+            ;;
+        # TEC_Idle, TEC_Watchdog
+        [2-3])  echo "Thormaster ($thorpid) Idle or unconnected to slave(s)"
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: unconnected => 'lost connection' would be better.

File: initfiles/componentfiles/thor/run_thor [hunk 5]
```diff
@@ -69,41 +69,65 @@ while [ 1 ]; do
...
+                # restarting thormaster
+                ;;
+        # TEC_SlaveInit
+        4)  echo "Thormaster ($thorpid) had slaves that failed to initialize"
+            echo "Shutting down"
+            $deploydir/stop_thor $deploydir
+            exit 1
+            ;;
+        # TEC_Swap
+        5)  echo "Thormaster ($thorpid) Thor Swap"
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: This is changing the autoSwapNode semantics.
TEC_Swap is new and an indication that Thor has detected the need/request to swap, however previously the autoswap was called after every Thor recycle, which performs additional checks to see if swap is warranted.

To TEC_Swap is a hint, but it shouldn't only perform the autoswap code if it gets that return code.

File: initfiles/componentfiles/thor/run_thor [hunk 6]
```diff
@@ -69,41 +69,65 @@ while [ 1 ]; do
...
+                $deploydir/stop_thor $deploydir
+                exit 1
+              fi
             fi
-        fi
+            ;;
+        # TEC_DaliDown
+        6)  echo "Thormaster ($thorpid) Unable to connect to Dali"
+            $deploydir/stop_thor $deploydir
+            exit 1
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: New behaviour, and I don't think desirable.
Previously Thor would keep recycling if Dali is down and when it is up, succeed.
Now it will exit and require manual intervention.

**Comment by jakesmith** [comment]
@Michael-Gardner - see comments.

File: initfiles/componentfiles/thor/run_thor (hunk 2 above)
Comment (Michael-Gardner) [inline initfiles/componentfiles/thor/run_thor]: The reason I kept this here was to insure pid file and sentinel file cleanup.  I'll go ahead and remove it and add sentinel file removal into the binary for when we exit with TEC_Clean (in HPCC-13156)

File: initfiles/componentfiles/thor/run_thor (hunk 3 above)
Comment (Michael-Gardner) [inline initfiles/componentfiles/thor/run_thor]: stop_thor uses the kill_process function from initfiles/sbin/hpcc_setenv.in.  I use it now for all the components so the binaries shut down in the same manner.  SIGTERM is what is being sent to the thormaster process.  Which is why HPCC-13156 exists.  Right now we're always seeing errors in the thormaster logs regarding shutdown.

**Comment by Michael-Gardner** [comment]
* PID file and uslaves.start cleanup when TEC_Clean is handled
* TEC_Clean from sigterm will remove sentinel file in binary https://track.hpccsystems.com/browse/HPCC-13437
* Moved TEC_Swap and TEC_DaliDown in with the same generic logic for TEC_Idle and TEC_Watchdog.  All 4 will do autoSwapNode if it is set, and attempt to restart thormaster.

Testing now.

File: initfiles/componentfiles/thor/run_thor (hunk 2 above)
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: I think it's fine as is, just questioning why different.
Clean stop should have involved stop_thor already, so normally this should not be necessary and be a NOP.

File: initfiles/componentfiles/thor/run_thor (hunk 3 above)
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: Ok, I see, makes sense.

File: initfiles/componentfiles/thor/run_thor [hunk 7]
```diff
@@ -69,41 +69,59 @@ while [ 1 ]; do
...
-            swapnode auto $DALISERVER $compname
-            errcode=$?
-            if [ 0 != ${errcode} ]; then
-                echo auto swap node failed, errcode=${errcode}
-            fi
-        fi
+            exit 1
+            ;;
+        # TEC_Idle, TEC_Watchdog, TEC_Swap, TEC_DaliDown
+        2|3|5|6)    [[ $errcode -lt 4 ]] && echo "Thormaster ($thorpid) Idle or lost connection to slave(s)"
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: Now common section, probably worth issuing a separate message her for TEC_Idle vs TEC_Watchdog

File: initfiles/componentfiles/thor/run_thor [hunk 8]
```diff
@@ -69,41 +69,59 @@ while [ 1 ]; do
...
+                    [[ $errcode -eq 5 ]] && echo "Thormaster ($thorpid) Exited with TEC_Swap"
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: Trivial: only message that mentions internal enum code, prob. best if says ".. Swap node required"

**Comment by jakesmith** [comment]
@Michael-Gardner - looks ok apart from minor comments re. messages.

**Comment by Michael-Gardner** [comment]
@jakesmith should be ready for a final look.

**Comment by jakesmith** [comment]
Looks ready to merge.
@richardkchapman
